File: system-test/java/src/main/java/com/optivem/shop/dsl/driver/adapter/external/erp/BaseErpDriver.java
package com.optivem.shop.dsl.driver.adapter.external.erp;

import com.optivem.shop.dsl.driver.port.external.erp.ErpDriver;

import com.optivem.shop.dsl.driver.adapter.external.erp.client.BaseErpClient;
import com.optivem.shop.dsl.driver.port.external.erp.dtos.GetProductRequest;
import com.optivem.shop.dsl.driver.port.external.erp.dtos.GetProductResponse;
import com.optivem.shop.dsl.driver.port.shared.dtos.ErrorResponse;
import com.optivem.shop.dsl.common.Closer;
import com.optivem.shop.dsl.common.Result;

public abstract class BaseErpDriver<TClient extends BaseErpClient> implements ErpDriver {
    protected final TClient client;

    protected BaseErpDriver(TClient client) {
        this.client = client;
    }

    @Override
    public void close() throws Exception {
        Closer.close(client);
    }

    @Override
    public Result<Void, ErrorResponse> goToErp() {
        return client.checkHealth()
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }

    @Override
    public Result<GetProductResponse, ErrorResponse> getProduct(GetProductRequest request) {
        return client.getProduct(request.getSku())
                .map(productDetails -> GetProductResponse.builder()
                        .sku(productDetails.getId())
                        .price(productDetails.getPrice())
                        .build())
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }
}

File: system-test/typescript/src/testkit/driver/adapter/external/erp/BaseErpDriver.ts
import type { Result } from '../../../../common/result.js';
import type { ErpErrorResponse } from '../../../port/external/erp/dtos/ErpErrorResponse.js';
import type { GetProductRequest } from '../../../port/external/erp/dtos/GetProductRequest.js';
import type { GetProductResponse } from '../../../port/external/erp/dtos/GetProductResponse.js';
import type { ReturnsProductRequest } from '../../../port/external/erp/dtos/ReturnsProductRequest.js';
import type { ReturnsPromotionRequest } from '../../../port/external/erp/dtos/ReturnsPromotionRequest.js';
import type { ErpDriver } from '../../../port/external/erp/erp-driver.js';
import type { BaseErpClient } from './client/BaseErpClient.js';

export abstract class BaseErpDriver<TClient extends BaseErpClient> implements ErpDriver {
  protected constructor(protected readonly client: TClient) {}

  async goToErp(): Promise<Result<void, ErpErrorResponse>> {
    return this.client.checkHealth();
  }

  async getProduct(request: GetProductRequest): Promise<Result<GetProductResponse, ErpErrorResponse>> {
    return this.client.getProduct(request.sku);
  }

  abstract returnsProduct(request: ReturnsProductRequest): Promise<Result<void, ErpErrorResponse>>;
  abstract returnsPromotion(request: ReturnsPromotionRequest): Promise<Result<void, ErpErrorResponse>>;
  abstract close(): Promise<void>;
}
